### PR TITLE
add topologyspreadconstraints policy

### PR DESCRIPTION
**Pull request by monotek**
Signed-off-by: André Bauer <andre.bauer@staffbase.com>

## Related Issue(s)
https://github.com/kyverno/policies/issues/429

## Description
- add topologyspreadconstraints policy

```
kyverno test .

Executing topologyspreadconstraints...
applying 1 policy to 5 resources... 

│───│─────────────│─────────────│──────────────────────────────│────────│
│ # │ POLICY      │ RULE        │ RESOURCE                     │ RESULT │
│───│─────────────│─────────────│──────────────────────────────│────────│
│ 1 │ spread-pods │ spread-pods │ monitoring/StatefulSet/pass  │ Pass   │
│ 2 │ spread-pods │ spread-pods │ monitoring/StatefulSet/fail1 │ Pass   │
│ 3 │ spread-pods │ spread-pods │ monitoring/StatefulSet/fail2 │ Pass   │
│ 4 │ spread-pods │ spread-pods │ monitoring/StatefulSet/fail3 │ Pass   │
│ 5 │ spread-pods │ spread-pods │ monitoring/StatefulSet/skip  │ Pass   │
│───│─────────────│─────────────│──────────────────────────────│────────│

Test Summary: 5 tests passed and 0 tests failed

```

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.

### CODE REVIEW

**Comment by monotek** [comment]
We might want to add a precondigtion here too:
```
      preconditions:
        all:
          - key: "{{ request.object.spec.replicas }}"
            operator: GreaterThanOrEquals
            value: 3
```

What do you think?

**Comment by chipzoller** [comment]
That'd probably be a good enhancement.

**Comment by monotek** [comment]
done

**Review by chipzoller — CHANGES_REQUESTED** [review]
Request you to rename the file `tsc-policy.yaml` to be more closely aligned with the policy name.

File: other/require-topologyspreadconstraints/tsc-policy.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: spread-pods
```
Comment (chipzoller) [inline other/require-topologyspreadconstraints/tsc-policy.yaml:4]: Name should be updated.

File: other/require-topologyspreadconstraints/tsc-policy.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  annotations:
```
Comment (chipzoller) [inline other/require-topologyspreadconstraints/tsc-policy.yaml:5]: Should be annotations for min Kyverno and min K8s versions.

File: other/require-topologyspreadconstraints/tsc-policy.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    policies.kyverno.io/title: Spread Pods Across Nodes & Zones
```
Comment (chipzoller) [inline other/require-topologyspreadconstraints/tsc-policy.yaml:6]: Title needs update

File: other/require-topologyspreadconstraints/tsc-policy.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    policies.kyverno.io/category: Sample
+    policies.kyverno.io/subject: Deployment, StatefulSet
+    policies.kyverno.io/description: >-
```
Comment (chipzoller) [inline other/require-topologyspreadconstraints/tsc-policy.yaml:9]: Description states 3 things:
1. What is the context
2. How does this policy add value
3. What does this policy do

File: other/require-topologyspreadconstraints/tsc-policy.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      Deployments to a Kubernetes cluster with multiple availability zones often need to
+      distribute those replicas to align with those zones to ensure site-level failures
+      do not impact availability. Therefore topologySpreadConstraints has to be defined.
+spec:
+  background: true
+  failurePolicy: Ignore
```
Comment (chipzoller) [inline other/require-topologyspreadconstraints/tsc-policy.yaml:15]: Remove please. We omit this in samples.

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Review by chipzoller — APPROVED** [review]

**Pull request merged**